samples/bluetooth/encrypted_advertising/peripheral/src/peripheral_ead.c

```diff
@@ -0,0 +1,377 @@
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	char passkey_str[7];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	snprintk(passkey_str, 7, "%06u", passkey);
+
+	LOG_DBG("Confirm passkey for %s: %s", addr, passkey_str);
+
+	k_poll_signal_raise(&passkey_enter_signal, 0);
+}
+
+static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	char passkey_str[7];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	snprintk(passkey_str, 7, "%06u", passkey);
+
+	LOG_DBG("Passkey for %s: %s", addr, passkey_str);
+}
+
+static void auth_cancel(struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("Pairing cancelled: %s", addr);
+}
+
+static int update_security(void)
+{
+	int err;
+
+	await_signal(&passkey_enter_signal);
+
+	err = bt_conn_auth_passkey_confirm(default_conn);
+	if (err) {
+		LOG_DBG("Failed to confirm passkey.");
+		return -1;
+	}
+
+	return 0;
+}
+
+static int init_bt(void)
+{
+	int err;
+
+	k_poll_signal_init(&disconn_signal);
+	k_poll_signal_init(&passkey_enter_signal);
+
+	err = bt_enable(NULL);
+	if (err) {
+		LOG_ERR("Bluetooth init failed (err %d)", err);
+		return -1;
+	}
+
+	LOG_DBG("Bluetooth initialized");
+
+	err = bt_unpair(BT_ID_DEFAULT, BT_ADDR_LE_ANY);
+	if (err) {
+		LOG_ERR("Unpairing failed (err %d)", err);
+	}
+
+	peripheral_cb.connected = connected;
+	peripheral_cb.disconnected = disconnected;
+	peripheral_cb.security_changed = security_changed;
+
+	bt_conn_cb_register(&peripheral_cb);
+
+	peripheral_auth_cb.pairing_confirm = NULL;
+	peripheral_auth_cb.passkey_confirm = auth_passkey_confirm;
+	peripheral_auth_cb.passkey_display = auth_passkey_display;
+	peripheral_auth_cb.passkey_entry = NULL;
+	peripheral_auth_cb.oob_data_request = NULL;
+	peripheral_auth_cb.cancel = auth_cancel;
+
+	err = bt_conn_auth_cb_register(&peripheral_auth_cb);
+	if (err) {
+		LOG_ERR("Failed to register bt_conn_auth_cb (err %d)", err);
+		return -1;
+	}
+
+	return 0;
+}
+
+int run_peripheral_sample(void)
+{
+	int err;
+	struct bt_le_ext_adv *adv = NULL;
+
+	err = init_bt();
+	if (err) {
+		return -1;
+	}
+
+	/* Setup advertiser */
+	err = create_adv(&adv);
+	if (err) {
+		return -1;
+	}
+	err = start_adv(adv);
+	if (err) {
+		return -1;
+	}
+	err = set_ad_data(adv);
+	if (err) {
+		return -1;
+	}
+
+	/* Wait for the peer to update security */
+	err = update_security();
+	if (err) {
+		LOG_ERR("Failure during security update");
+		return -1;
+	}
+
+	/* Wait for the peer to disconnect */
+	await_signal(&disconn_signal);
+
+	/* Restart advertising */
+	err = start_adv(adv);
+	if (err) {
+		return -1;
+	}
+	err = set_ad_data(adv);
+	if (err) {
+		return -1;
+	}
+
+	/* Wait 10s before stopping and deleting the advertiser */
+	k_sleep(K_SECONDS(10));
+
+	err = stop_and_delete_adv(adv);
+	if (err) {
+		return -1;
+	}
```

Comment:
Any reason why you are not returning `err` instead of `-1`? 

Should we also not log the error?

Reply:
The error code are meaningless here. This is user application, there is no upper layer to handle the error. They are here for the test to fail.

The logging of error is done in function, there is less code duplication like that.

Reply:
> This is user application, there is no upper layer to handle the error. 

Even in this PR you have an application (your babblesim test) that calls this, so there is in fact an upper layer to handle the error.

Reply:
Right but the test will not do anything special based on the error code. 

And if I return the error code from two different functions, they may return the same value but with a different meaning. 

If you really think that I should return special error code then I guess I should return a different value for each different functions.

Reply:
> Right but the test will not do anything special based on the error code

_yet_ :) 

> If you really think that I should return special error code then I guess I should return a different value for each different functions.

I think we should return the error code. It doesn't necessarily need to be a unique error code per condition, but it would make sense to actually follow the way we return error codes in Zephyr